### PR TITLE
Introduce SWIFT_EXTERN_RUNTIME_EXPORT for Windows

### PR DESCRIPTION
**Pull request by hughbe**
IRGen includes Metdata.h. This means that Metadata.h has to be able to be compiled with MSVC, even though its part of the runtime.

The correct syntax for dll exporting an extern C symbol is

`extern "C" __declspec(dllexport_) ...`

However, current we do

`__declspec(dllexport) ... extern "C"`

This means that we get several hundred MSVC errors compiling IRGen.

The fix is to introduce a macro, `SWIFT_EXTERN_RUNTIME_EXPORT` that fixes the `declspec` errors.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: include/swift/Runtime/Config.h
Comment (compnerd) [inline include/swift/Runtime/Config.h]: I think this should be `SWIFT_RT_EXTERN_ENTRY_VISIBILITY` at the very least.

Is there ever a case we wouldnt want the `extern "C"` on the interface?

File: stdlib/public/SwiftShims/Visibility.h
Comment (compnerd) [inline stdlib/public/SwiftShims/Visibility.h]: This should be unnecessary.  Use the form from `_MSC_VER` clause always.

**Review by compnerd — DISMISSED** [review]

File: include/swift/Runtime/Config.h
Comment (hughbe) [inline include/swift/Runtime/Config.h]: There are two such situations in Casting.cpp
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Casting.cpp#L266
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Casting.cpp#L2423

As well as in Enum.cpp

- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Enum.cpp#L203
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Enum.cpp#L2

As well as Heap.cpp

- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Heap.cpp#L25
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Heap.cpp#L34

As well as HeapObject.cpp
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Heap.cpp#L34
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L307
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L330
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L386
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L402
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L410
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L415
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L432
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L463
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L475
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/HeapObject.cpp#L605

As well as Metadata.cpp
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Metadata.cpp#L230
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/Metadata.cpp#L2364

As well as SwiftObject.mm
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/SwiftObject.mm#L1345
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/SwiftObject.mm#L1444
- https://github.com/apple/swift/blob/master/stdlib/public/runtime/SwiftObject.mm#L1455

Comment (hughbe) [inline include/swift/Runtime/Config.h]: Judging by the number of these, I say we keep these two separate? I'll rename to SWIFT_RT...

File: stdlib/public/SwiftShims/Visibility.h
Comment (hughbe) [inline stdlib/public/SwiftShims/Visibility.h]: Sure, fixed

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

File: include/swift/Runtime/Config.h
Comment (gparker42) [inline include/swift/Runtime/Config.h]: Every export from the Swift runtime ought to be `extern "C"`.

Comment (hughbe) [inline include/swift/Runtime/Config.h]: Hmm, interesting only cpp files have export without extern - maybe this is an oversight.

Do we make the cpp files use extern C, or simply remove them - not sure if the headers are already annotated as extern C

Comment (gparker42) [inline include/swift/Runtime/Config.h]: I think we want this:
1. `SWIFT_RUNTIME_EXPORT` has both `extern "C"` and `__declspec` as appropriate
2. every export has a header declaration
3. header declarations of exports all use `SWIFT_RUNTIME_EXPORT`
4. non-header definitions do not use `SWIFT_RUNTIME_EXPORT` or `extern "C"` or `__declspec`.

Comment (hughbe) [inline include/swift/Runtime/Config.h]: Thanks. I'll update this PR accordingly.


What about the case of `extern "C"` without `SWIFT_RUNTIME_EXPORT`, such as in https://github.com/apple/swift/blob/adc54c8a4d13fbebfeb68244bac401ef2528d6d0/stdlib/public/runtime/SwiftHashableSupport.h

Comment (gparker42) [inline include/swift/Runtime/Config.h]: Those should remain simply `extern "C"`. Those exist simply to make the symbol name match the @_silgen_name used on the Swift side. They aren't exported anywhere so `__declspec` doesn't apply.

Comment (hughbe) [inline include/swift/Runtime/Config.h]: I see, cheers. I assume this is because `extern "C"` means that the symbol name stays the same as the name of the method - i.e. not mangled. Do you think we need to make this a bit more explicit and self documenting, via some macro (`#define SWIFT_NON_MANGLED_SYMBOL extern "C"`) or something like that? 
Might not be worth it as probably most people digging into the runtime (apart from me) understand C linkages well. On the other hand, would make out intentions explicit

Comment (gparker42) [inline include/swift/Runtime/Config.h]: I think a better way to make intentions explicit there would be to have a `SWIFT_CALLED_VIA_SILGEN_NAME` marker. That would incorporate `extern "C"` and potentially be useful for keeping the C++ side and the Swift side in sync.

Comment (gparker42) [inline include/swift/Runtime/Config.h]: …but IMO that ought to be a job for a follow-up PR.

Comment (hughbe) [inline include/swift/Runtime/Config.h]: Yeah I like that. I'll submit a PR once I get the other extern c stuff merged :)

Cheers

Comment (hughbe) [inline include/swift/Runtime/Config.h]: > …but IMO that ought to be a job for a follow-up PR.

Took the words out of my mouth ;)

Comment (gparker42) [inline include/swift/Runtime/Config.h]: Great minds think alike, obviously.

**Comment by hughbe** [comment]
Okay, I've updated this PR to do the following:
- Define SWIFT_RUNTIME_EXPORT in terms of extern "C"
- Remove all references to extern "C" if we have SWIFT_RUNTIME_EXPORT/SWIFT_RUNTIME_STDLIB_INTERFACE/SWIFT_RT_ENTRY_VISIBILITY

I've tested on Windows, everything works, but I'll trigger on Linux/OSX (for objc only methods) before the review

I'm attaching a summary of my work as well - it details examples where we used one of these macros without extern "C". Most of these will be cleaned up in a future PR

[SWIFT_RT_ENTRY_VISIBILITY.txt](https://github.com/apple/swift/files/719679/SWIFT_RT_ENTRY_VISIBILITY.txt)
[SWIFT_RUNTIME_EXPORT.txt](https://github.com/apple/swift/files/719681/SWIFT_RUNTIME_EXPORT.txt)
[SWIFT_RUNTIME_STDLIB_INTERFACE.txt](https://github.com/apple/swift/files/719680/SWIFT_RUNTIME_STDLIB_INTERFACE.txt)

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by hughbe** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@hughbe Sorry, for now only those with commit access can trigger CI.

**Comment by slavapestov** [comment]
Oh, well then CI was just being flaky.

**Comment by hughbe** [comment]
@swift-ci please test

**Comment by slavapestov** [comment]
@compnerd Mind clicking the big green button? :)

**Comment by compnerd** [comment]
Yeah, given the follow up work and the changes now, LG.

**Comment by gparker42** [comment]
Thanks, Hugh.

**Pull request merged**
